velox/dwio/text/writer/TextWriter.cpp

```diff
@@ -0,0 +1,142 @@
+    case TypeKind::ARRAY:
+      [[fallthrough]];
+    case TypeKind::MAP:
+      [[fallthrough]];
+    case TypeKind::ROW:
+      [[fallthrough]];
+    case TypeKind::UNKNOWN:
+      [[fallthrough]];
+    default:
+      VELOX_NYI("{} is not supported yet in TextWriter", type->kind());
+  }
+}
+
+TextWriter::TextWriter(
+    RowTypePtr schema,
+    std::unique_ptr<dwio::common::FileSink> sink,
+    const std::shared_ptr<text::WriterOptions>& options)
+    : schema_(std::move(schema)) {
+  bufferedWriterSink_ = std::make_unique<BufferedWriterSink>(
+      std::move(sink),
+      options->memoryPool->addLeafChild(fmt::format(
+          "{}.text_writer_node.{}",
+          options->memoryPool->name(),
+          folly::to<std::string>(folly::Random::rand64()))),
+      options->defaultFlushCount);
+}
+
+void TextWriter::write(const VectorPtr& data) {
+  VELOX_CHECK_EQ(
+      data->encoding(),
+      VectorEncoding::Simple::ROW,
+      "Text writer expects row vector input");
+  VELOX_CHECK(
+      data->type()->equivalent(*schema_),
+      "The file schema type should be equal with the input row vector type.");
+  const RowVector* dataRowVector = data->as<RowVector>();
+
+  std::vector<std::shared_ptr<DecodedVector>> decodedColumnVectors;
+  const auto numColumns = dataRowVector->childrenSize();
+  for (size_t column = 0; column < numColumns; ++column) {
+    auto decodedColumnVector = std::make_shared<DecodedVector>(DecodedVector(
+        *dataRowVector->childAt(column),
+        SelectivityVector(dataRowVector->size())));
+    decodedColumnVectors.push_back(decodedColumnVector);
+  }
+
+  for (vector_size_t row = 0; row < data->size(); ++row) {
+    for (size_t column = 0; column < numColumns; ++column) {
+      if (column != 0) {
+        bufferedWriterSink_->write(SOH);
+      }
+      auto columnData = encodeColumnCellValue(
+          decodedColumnVectors.at(column), schema_->childAt(column), row);
+      bufferedWriterSink_->write(columnData.c_str(), columnData.length());
+    }
+    bufferedWriterSink_->write(newLine);
+  }
+}
+
+void TextWriter::flush() {
+  bufferedWriterSink_->flush();
+}
+
+void TextWriter::close() {
+  bufferedWriterSink_->close();
+}
+
+void TextWriter::abort() {
+  bufferedWriterSink_->abort();
+}
+
+std::unique_ptr<dwio::common::Writer> TextWriterFactory::createWriter(
+    std::unique_ptr<dwio::common::FileSink> sink,
+    const std::shared_ptr<dwio::common::WriterOptions>& options) {
+  auto textOptions = std::dynamic_pointer_cast<text::WriterOptions>(options);
+  VELOX_CHECK_NOT_NULL(
+      textOptions, "Text writer factory expected a Text WriterOptions object.");
+  return std::make_unique<TextWriter>(
+      asRowType(options->schema), std::move(sink), textOptions);
+}
+
+std::unique_ptr<dwio::common::WriterOptions>
+TextWriterFactory::createWriterOptions() {
+  return std::make_unique<text::WriterOptions>();
+}
+
+const std::string TextWriter::nullData = "\\N";
```

Comment:
static inline std::string nullData (in header) then we don't need to declare in cc.